scrapy_poet/injection.py

```diff
@@ -0,0 +1,323 @@
+        for cls, kwargs_spec in plan.dependencies:
+            if cls not in instances.keys():
+                instances[cls] = cls(**kwargs_spec.kwargs(instances))
+
+        raise returnValue(instances)
+
+    @inlineCallbacks
+    def build_instances_from_providers(
+            self, request: Request, response: Response, plan: andi.Plan):
+        """Build dependencies handled by registered providers"""
+        instances: Dict[Callable, Any] = {}
+        scrapy_provided_dependencies = self.available_dependencies_for_providers(
+            request, response)
+        dependencies_set = {cls for cls, _ in plan.dependencies}
+        for provider in self.providers:
+            provided_classes = {cls for cls in dependencies_set if
+                                provider.is_provided(cls)}
+            provided_classes -= instances.keys()  # ignore already provided types
+            if not provided_classes:
+                continue
+
+            kwargs = andi.plan(
+                provider,
+                is_injectable=is_injectable,
+                externally_provided=scrapy_provided_dependencies,
+                full_final_kwargs=False,
+            ).final_kwargs(scrapy_provided_dependencies)
+            results = yield maybeDeferred_coro(provider, set(provided_classes),
+                                               **kwargs)
+            extra_classes = results.keys() - provided_classes
+            if extra_classes:
+                raise UndeclaredProvidedTypeError(
+                    f"{provider} has returned {extra_classes} but they're not "
+                    f"listed as provided classes {provider.provided_classes}"
+                )
+            instances.update(results)
+
+        raise returnValue(instances)
+
+    @inlineCallbacks
+    def build_callback_dependencies(self, request: Request, response: Response):
+        """
+        Scan the configured callback for this request looking for the
+        dependencies and build the corresponding instances. Return a kwargs
+        dictionary with the built instances.
+        """
+        plan = self.build_plan(request)
+        provider_instances = yield from self.build_instances(request, response, plan)
+        callback_kwargs = plan.final_kwargs(provider_instances)
+        raise returnValue(callback_kwargs)
+
+
+def check_all_providers_are_callable(providers):
+    for provider in providers:
+        if not callable(provider):
+            raise NonCallableProviderError(
+                f"The provider {type(provider)} is not callable. "
+                f"It must implement '__call__' method"
+            )
+
+
+def load_provider_classes(settings: Settings) -> List[PageObjectInputProvider]:
+    providers = settings.getlist('SCRAPY_POET_PROVIDER_CLASSES') or PROVIDERS
+    logger.info(f"Loading providers {providers}")
+    result = []
+    for cls in settings.getlist('SCRAPY_POET_PROVIDER_CLASSES') or PROVIDERS:
+        if not callable(cls):
```

Comment:
FTR, this check is not needed since scrapy 2.4

Reply:
Removed